### PR TITLE
[CDAP-12423] Refactor AppMetadataStore to prevent program status change race condition

### PR DESCRIPTION
**Pull request by ecfm**
https://issues.cask.co/browse/CDAP-12423

Design: https://wiki.cask.co/display/CE/Program+State+Transition+for+4.3

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 1]
```diff
@@ -321,170 +338,202 @@ private void recordProgramStart(ProgramId programId, String pid, long startTs, S
...
+    if (!isValid && !existingRecords.isEmpty()) {
+      // Skip recording start if the existing records are not valid
+      return;
+    }
+    RunRecordMeta existing = null;
+    // Only if existingRecords is empty & upgrade is not complete & the version is default version,
+    // also try to get the record without the version string
+    if (existingRecords.isEmpty()) {
+      if (!upgradeComplete.get() && programId.getVersion().equals(ApplicationId.DEFAULT_VERSION)) {
+        MDSKey key = getVersionLessProgramKeyBuilder(TYPE_RUN_RECORD_SUSPENDED, programId)
```
Comment (ecfm) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java:475]: Copied this from existing code. However, we actually only upgrade record of type `TYPE_RUN_RECORD_COMPLETED` with the assumption all programs are stopped before upgrade. However, here the assumption is that there are still records of type `TYPE_RUN_RECORD_SUSPENDED`. We should only have one assumption. Still asking Gokul for confirmation

**Comment by anew** [comment]
@maochf it would be nice if you explain what the race condition is, in the PR description. I am not sure I understand what we are fixing here

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 2]
```diff
@@ -282,14 +283,14 @@ private void addWorkflowNodeState(ProgramId programId, String pid, Map<String, S
...
       Map<String, String> properties = record.getProperties();
       properties.put(workflowNodeId, pid);
-      write(key, new RunRecordMeta(record, properties));
+      write(key, new RunRecordMeta(record, properties, sourceId));
     }
   }
 
   public void recordProgramStart(ProgramId programId, String pid, long startTs, String twillRunId,
-                                 Map<String, String> runtimeArgs, Map<String, String> systemArgs) {
+                                 Map<String, String> runtimeArgs, Map<String, String> systemArgs, byte[] sourceId) {
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: what is a source ID? This is a public method, why is there no Javadoc?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 3]
```diff
@@ -305,11 +306,29 @@ void recordProgramRunningOldFormat(ProgramId programId, String pid, long stateCh
...
+                                  MDSKey.Builder keyBuilder, byte[] sourceId) {
+    List<RunRecordMeta> existingRecords = getRuns(programId, pid);
+    if (existingRecords.size() > 1) {
+      // This shouold never happen
+      LOG.warn("Found more than 1 existing run records '{}' for program '{}' with run id '{}'. " +
+                 "Skip recording program start for it.", existingRecords, programId, pid);
+      return;
+    }
+    if (existingRecords.size() == 1) {
+      byte[] existingSourceId = existingRecords.get(0).getSourceId();
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: how can this happen?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 4]
```diff
@@ -418,7 +438,7 @@ private void recordProgramSuspendResume(ProgramId programId, String pid, String
     key = getProgramKeyBuilder(toType, programId)
       .add(pid)
       .build();
-    write(key, new RunRecordMeta(record, null, toStatus));
+    write(key, new RunRecordMeta(record, null, toStatus, ));
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: extra comma?

Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: , sourceId?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java [hunk 5]
```diff
@@ -214,7 +214,7 @@ public Boolean call(DatasetContext context) throws Exception {
               if (systemArgs == null) {
                 systemArgs = EMPTY_STRING_MAP;
               }
-              mds.recordProgramStart(id, pid, nowSecs, target.getTwillRunId(), runtimeArgs, systemArgs);
+              mds.recordProgramStart(id, pid, nowSecs, target.getTwillRunId(), runtimeArgs, systemArgs, messageId.getRawId());
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java]: so, a sourceId is a message Id?

**Comment by ecfm** [comment]
@anew One existing race condition is between `RunrecordCorrector` and `ProgramNotificationSubsriberService`. When `RunrecordCorrector` corrects a running record, the notification of program run status KILLED or FAILED may have already be published, but `ProgramNotificationSubsriberService` has not yet processed. With this change `RunrecordCorrector` will also publish notification to fix run record, so that the published notification will be processed first.

In addition, the messageId is persisted as part of run record, and when persisting new run records, if the existing run record contains a messageId larger than that of the new run record, the new run record will be ignored. Therefore we can replay messages to get reprocessed by  `ProgramNotificationSubsriberService`.

The major changes in this PR are recording run record according to the state diagram in this design: https://wiki.cask.co/display/CE/Program+State+Transition+for+4.3
The states represent the type of run record, and the actions on the arrows are the recording methods in `AppMetaStore`. Actions in black are legal state transitions, actions in red are still allowed but should be logged. Self-pointing arrows are also allowed but should be logged. The IGNORE state represent the current run record should not be persisted and should log as error

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 6]
```diff
@@ -346,155 +344,126 @@ private void recordProgramStart(ProgramId programId, String pid, long startTs, S
...
-        String msg = String.format("No meta for run record for namespace %s app %s program type %s " +
-                                   "program %s pid %s exists",
-                                   programId.getNamespace(), programId.getApplication(), programId.getType().name(),
-                                   programId.getProgram(), pid);
-        LOG.error(msg);
-        throw new IllegalArgumentException(msg);
-      }
+                                    MDSKey.Builder keyBuilder, byte[] sourceId) {
+    List<RunRecordMeta> existingRecords = getRuns(programId, pid);
+    boolean isValid = validateExistingRecords(existingRecords, programId, pid, sourceId, "start",
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: should this be "running" instead of "start"?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 7]
```diff
@@ -346,155 +344,126 @@ private void recordProgramStart(ProgramId programId, String pid, long startTs, S
...
+                                              ImmutableSet.of(ProgramRunStatus.STARTING),
+                                              ImmutableSet.of(ProgramRunStatus.SUSPENDED, ProgramRunStatus.RUNNING));
+    if (!isValid) {
+      // Skip recording start if the existing records are not valid
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: start/running

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 8]
```diff
@@ -346,155 +344,126 @@ private void recordProgramStart(ProgramId programId, String pid, long startTs, S
...
 
-  public void recordProgramSuspend(ProgramId program, String pid) {
-    recordProgramSuspendResume(program, pid, "suspend");
+  public void recordProgramSuspend(ProgramId programId, String pid, byte[] sourceId) {
+    List<RunRecordMeta> existingRecords = getRuns(programId, pid);
+    boolean isValid = validateExistingRecords(existingRecords, programId, pid, sourceId, "suspend",
+                                              ImmutableSet.of(ProgramRunStatus.STARTING, ProgramRunStatus.RUNNING),
+                                              ImmutableSet.of(ProgramRunStatus.SUSPENDED));
+    if (!isValid) {
+      // Skip recording start if the existing records are not valid
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: start/suspend

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 9]
```diff
@@ -346,155 +344,126 @@ private void recordProgramStart(ProgramId programId, String pid, long startTs, S
...
+    // Only one record in the valid existing records
+    RunRecordMeta existing = existingRecords.get(0);
+    recordProgramSuspendResume(programId, pid, sourceId, existing, "suspend");
   }
 
-  public void recordProgramResumed(ProgramId program, String pid) {
-    recordProgramSuspendResume(program, pid, "resume");
+  public void recordProgramResumed(ProgramId programId, String pid, byte[] sourceId) {
+    List<RunRecordMeta> existingRecords = getRuns(programId, pid);
+    boolean isValid = validateExistingRecords(existingRecords, programId, pid, sourceId, "suspend",
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: suspend/resumed

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 10]
```diff
@@ -346,155 +344,126 @@ private void recordProgramStart(ProgramId programId, String pid, long startTs, S
...
+                                              ImmutableSet.of(ProgramRunStatus.SUSPENDED),
+                                              ImmutableSet.of(ProgramRunStatus.STARTING, ProgramRunStatus.RUNNING));
+    if (!isValid && !existingRecords.isEmpty()) {
+      // Skip recording start if the existing records are not valid
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: start/resumed

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 11]
```diff
@@ -517,6 +486,51 @@ private void recordProgramStop(ProgramId programId, String pid, long stopTs, Pro
     return get(keySet, RunRecordMeta.class);
   }
 
+  private boolean validateExistingRecords(List<RunRecordMeta> existingRecords, ProgramId programId, String pid,
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: this needs a Javadoc. To explain what allowedStatuses means vs. allowedStatusesWithLogStatuses

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java [hunk 12]
```diff
@@ -217,18 +217,18 @@ public Boolean call(DatasetContext context) throws Exception {
               mds.recordProgramStart(id, pid, nowSecs, target.getTwillRunId(), runtimeArgs, systemArgs, messageId.getRawId());
               break;
             case RUNNING:
-              mds.recordProgramRunning(id, pid, nowSecs, target.getTwillRunId());
+              mds.recordProgramRunning(id, pid, nowSecs, target.getTwillRunId(), );
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java]: the comma without another parameter looks like a syntax error... why are you doing this?

**Comment by anew** [comment]
Ah that makes sense. I guess the state diagram is represented by all the method calls to validateExistingRecords() in AppMetaDataStore. Is there a way we could make that more explicit? Like, store the transition diagram in a table and use that to determine what states are valid?

**Review by anew — COMMENTED** [review]
Overall, these changes makes sense. Please see my comments.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java [hunk 13]
```diff
@@ -56,64 +55,70 @@
...
    */
   void setStart(ProgramId id, String pid, long startTime, @Nullable String twillRunId,
-                Map<String, String> runtimeArgs, Map<String, String> systemArgs);
+                Map<String, String> runtimeArgs, Map<String, String> systemArgs, byte[] sourceId);
 
   /**
    * Logs start of program run and persists program status to {@link ProgramRunStatus#RUNNING}.
-   *
-   * @param id id of the program
+   *  @param id id of the program
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java]: fix indentation.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java [hunk 14]
```diff
@@ -56,64 +55,70 @@
...
    */
-  void setRunning(ProgramId id, String pid, long runTime, @Nullable String twillRunId);
+  void setRunning(ProgramId id, String pid, long runTime, @Nullable String twillRunId, byte[] sourceId);
 
   /**
    * Logs end of program run and sets the run status to one of: {@link ProgramRunStatus#COMPLETED},
    * or {@link ProgramRunStatus#KILLED}.
-   *
-   * @param id id of the program
+   *  @param id id of the program
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java]: indentation

Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java]: also several times below. please fix all.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java (hunk 13 above)
Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java]: Also, you should not delete the empty line between the description and the @param's (per Javadoc style convention)

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 15]
```diff
@@ -501,11 +501,16 @@ private boolean validateExistingRecords(List<RunRecordMeta> existingRecords, Pro
...
     }
-    if (existingRecords.isEmpty()) {
-      // If existing record is not allowed, only empty existing records is valid
-      if (allowedStatuses.isEmpty() && allowedWithLogStatuses.isEmpty()) {
+    // If existing record is not allowed, only empty existing records is valid
+    if (allowedStatuses.isEmpty() && allowedWithLogStatuses.isEmpty()) {
+      if (existingRecords.isEmpty()) {
         return true;
       }
+      LOG.error("No existing run record meta should exist for program '{}' pid '{}' but found '{}'.",
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: what is a pid? Is that a run ID? Then that is what the message should say.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 16]
```diff
@@ -501,11 +501,16 @@ private boolean validateExistingRecords(List<RunRecordMeta> existingRecords, Pro
...
+                programId, pid, recordType, existingRecords);
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: the message is odd. It has more arguments than placeholders.

Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: I think the messages need some work, so that our users can also understand them. Let's do that in a follow-up PR.

File: cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/log/MockLogReader.java [hunk 17]
```diff
@@ -77,12 +78,28 @@
...
   private final List<LogEvent> logEvents = Lists.newArrayList();
   private final Map<ProgramId, RunRecord> runRecordMap = Maps.newHashMap();
+  private int sourceId; 
 
   @Inject
   MockLogReader(DefaultStore store) {
     this.store = store;
   }
 
+  private void setStartAndRunning(final ProgramId id, final String pid, final long startTime) {
```
Comment (anew) [inline cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/log/MockLogReader.java:88]: +1 for moving these from the main to the test classes. But is MockLogReader the right place for these methods?

Comment (anew) [reply cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/log/MockLogReader.java:88]: oh is this the only place where they are needed? Then it's ok.

File: cdap-app-fabric-tests/src/test/java/co/cask/cdap/internal/app/services/http/handlers/WorkflowStatsSLAHttpHandlerTest.java [hunk 18]
```diff
@@ -63,13 +65,30 @@
...
+  private int sourceId;
+
   @BeforeClass
   public static void beforeClass() throws Throwable {
     AppFabricTestBase.beforeClass();
     store = getInjector().getInstance(DefaultStore.class);
     metricStore = getInjector().getInstance(MetricStore.class);
   }
 
+  private void setStartAndRunning(final ProgramId id, final String pid, final long startTime) {
```
Comment (anew) [inline cdap-app-fabric-tests/src/test/java/co/cask/cdap/internal/app/services/http/handlers/WorkflowStatsSLAHttpHandlerTest.java:76]: instead of copying these to every test case that needs them, it may be better to 
- either add  them to AppFabricTestBase (if all tests extend from that)
- or create a class with two static methods that can be called from all tests

File: cdap-app-fabric-tests/src/test/java/co/cask/cdap/internal/app/services/http/handlers/WorkflowStatsSLAHttpHandlerTest.java [hunk 19]
```diff
@@ -63,13 +65,30 @@
...
+
+  }
+
+
+  private void setStartAndRunning(final ProgramId id, final String pid, final long startTime,
+                                  final Map<String, String> runtimeArgs,
+                                  final Map<String, String> systemArgs) {
+    store.setStart(id, pid, startTime, null, runtimeArgs, systemArgs,
+                   ByteBuffer.allocate(++sourceId).array());
+    store.setRunning(id, pid, startTime + 1, null, ByteBuffer.allocate(++sourceId).array());
```
Comment (anew) [inline cdap-app-fabric-tests/src/test/java/co/cask/cdap/internal/app/services/http/handlers/WorkflowStatsSLAHttpHandlerTest.java]: ByteBuffer.allocate(++sourceId).array() ? Is that really what you want? Create a byte buffer with capacity of sourceId, and then get the backing array? That array will not be exactly as long as a messageId. and it will be filled with zeros always. 

If I understand correctly, what you really need is to generate message Ids with increasing publish time. You can do that as follows:
```
static long sourceId = System.currentTimeMillis();
static byte[] createSourceMessageId() {
  byte[] buffer = new byte[MessageId.RAW_ID_SIZE];
  messageId.putRaw(sourceId++, 0, 0, 0, buffer, 0);
  return buffer;
}
```
By the way, you should never use array() on a ByteBuffer. It is easily perceived as returning the byte[] representation of the ByteBuffer, but that is not the case.

Comment (anew) [reply cdap-app-fabric-tests/src/test/java/co/cask/cdap/internal/app/services/http/handlers/WorkflowStatsSLAHttpHandlerTest.java]: Similar to the other two methods, if you place this as a static method in a class that all tests can access, you don't need to copy it to every test case.

**Review by anew — COMMENTED** [review]
Please see my comments. In addition:
- are you going to add test case for out-of-order source ids?
- are you going to add test cases for transitions that are not allowed and ignored?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java (hunk 3 above)
Comment (ecfm) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: which case are you referring to?

Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: I think I meant the case where there are more than 1. Must have clicked the wrong line.

Comment (ecfm) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: that's just a sanity check and should never happen

**Comment by anew** [comment]
Last two commits look good, only thing left is test refactoring

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/AppMetadataStoreTest.java [hunk 20]
```diff
@@ -134,9 +130,8 @@ public void apply() throws Exception {
     });
   }
 
-  @Ignore // Newly added; still debugging
```
Comment (poornachandra) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/AppMetadataStoreTest.java]: Are we going to un-ignore this test in this PR? Looks like it needs more changes. Maybe you can do it in a separate PR. Here is the JIRA to track it - https://issues.cask.co/browse/CDAP-12492

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/RunRecordCorrectorServiceTest.java [hunk 21]
```diff
@@ -129,6 +129,8 @@ public ProgramRunStatus call() throws Exception {
     // Start the RunRecordCorrectorService, which will fix the run record
     new LocalRunRecordCorrectorService(store, programStateWriter, programLifecycleService, runtimeService).startUp();
 
+    // Sleep to make sure the run record correction notification is processed
+    Thread.sleep(3000);
```
Comment (poornachandra) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/RunRecordCorrectorServiceTest.java]: Let's use `Tasks.waitFor()` with 30 seconds, so that test does not become flaky.

**Review by poornachandra — COMMENTED** [review]
@maochf Just a couple of comments.
